### DISCUSSION TITLE
Kagi Search Error

### DISCUSSION THREAD

**Discussion by freeload101** [open-webui/open-webui#8132] [Issues]
# Bug Report
 
<3 all the work on OpenWeb UI !!

## Installation Method
https://github.com/freeload101/SCRIPTS/blob/master/Bash/OpenWebUI_Fast.bash

## Environment

**Open WebUI Version:** 
 ghcr.io/open-webui/open-webui:dev-cuda
 
**Ollama (if applicable):** 
ollama version is 0.5.4

**Operating System:**
OS Name	Microsoft Windows 11 Pro	
Version	10.0.22621 Build 22621	

**Browser (if applicable):** [e.g., Chrome 100.0, Firefox 98.0]
Version 126.0.6478.115 (Official Build) (64-bit)

**Confirmation:**

- [X] I have read and followed all the instructions provided in the README.md.
- [X] I am on the latest version of both Open WebUI and Ollama.
- [X] I have included the browser console logs.
- [X] I have included the Docker container logs.
- [X] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:
Kagi search work like google search works. even with google errors I can see the API key it uses but with Kagi I don't see it in the error

## Actual Behavior:
Red error box

## Description

**Bug Summary:**
Google search works but Kagi does not. Do I need the invite only API key or does it work with FAST api ? I have 25$ in credits on my account

## Reproduction Details

**Steps to Reproduce:**
Invoke any search 

## Logs and Screenshots
![{6C055F68-2F78-462F-A875-A06DCABB43A2}](https://github.com/user-attachments/assets/33824d42-f06f-46ac-8c83-2b5a668f83f3)

**Browser Console Logs:** 
[localhost_chrome_logs.har.txt](https://github.com/user-attachments/files/18214868/localhost_chrome_logs.har.txt)

**Docker Container Logs:**
```
ERROR [open_webui.routers.retrieval] 401 Client Error: Unauthorized for url: https://kagi.com/api/v0/search?q=weather+forecast+for+today&limit=6
Traceback (most recent call last):
  File "/app/backend/open_webui/routers/retrieval.py", line 1248, in process_web_search
    web_results = search_web(
                  ^^^^^^^^^^^
  File "/app/backend/open_webui/routers/retrieval.py", line 1146, in search_web
    return search_kagi(
           ^^^^^^^^^^^^
  File "/app/backend/open_webui/retrieval/web/kagi.py", line 31, in search_kagi
    response.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://kagi.com/api/v0/search?q=weather+forecast+for+today&limit=6
INFO:     47.4.236.77:0 - "POST /api/v1/retrieval/process/web/search HTTP/1.1" 400 Bad Request
INFO:     47.4.236.77:0 - "POST /api/chat/completions HTTP/1.1" 200 OK
INFO:     47.4.236.77:0 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO  [open_webui.routers.openai] get_all_models()
INFO:     47.4.236.77:0 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO  [open_webui.routers.ollama] get_all_models()
Count of chats for tag 'general': 0
**Screenshots/Screen Recordings (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]
```

## Additional Information
If you need a key hit me up rmccurdy.com

**Comment by tjbck** [comment]
@erics118

**Comment by erics118** [comment]
See https://help.kagi.com/kagi/api/search.html The fastgpt api key will not work.

**Reply by freeload101** [reply]
Yup. I haven't looked at writing my own API .. against TOS and out of scope for OpenWebUI but ;)

```
{"meta":{"id":"20e6de3156a023e937ff283bcfc9a0e5","node":"us-east4","ms":26,"api_balance":28.51},"data":null,"error":[{"code":2,"msg":"Unauthorized: the Search API is in closed beta and is currently available to Kagi Business (Team) plan.","ref":null}]}r
```

**Reply by wsargent** [reply]
For me it says:

```
"error":[{"code":2,"msg":"Unauthorized: Search API is currently in beta. Email support@kagi.com to request access.","ref":null}]}%                                                                                    
```

**Comment by legolije** [comment]
I'm not sure why an API key is offered if it's not possible to use it. The volume it'd be used at through a local instance of open-webui is negligible, probably no more than used through the search.

**Comment by 8bitbuddhist** [comment]
I had this same issue, but my problem was that the Search API is in a closed beta, and you have to request access from their support team. I emailed them requesting access, and they approved it within the hour. Then, you can [create your API key](https://kagi.com/settings?p=api) and it should work.

**Reply by solclam** [reply]
same here just bumping this for others since its incredibly confusing that you CAN create an API key in the settings menu but you cant actually use it unless you email and they activate it for you haha   edit: spelling

**Comment by panzer-arc** [comment]
You can have an active API but also be out of credits. The OWUI logs don't warn you that the lack of credits is the issue. Make sure you have enough in the account to pay for the queries, too.

**Comment by joey-zwaan** [comment]
As of today you still need to request access!

**Comment by russelg** [comment]
Even with access now granted, this doesn't work. Seems openwebui wants to use API v0, which doesn't seem to accept any of the access tokens I have, neither personal ones from https://kagi.com/settings/api nor the new ones from https://kagi.com/api/keys

For now, here's an edited `kagi.py` that supports the v1 API.

```python
import logging
from typing import Optional

import requests
from open_webui.retrieval.web.main import SearchResult, get_filtered_results

log = logging.getLogger(__name__)


def search_kagi(api_key: str, query: str, count: int, filter_list: Optional[list[str]] = None) -> list[SearchResult]:
    """Search using Kagi's Search API and return the results as a list of SearchResult objects.

    The Search API will inherit the settings in your account, including results personalization and snippet length.

    Args:
        api_key (str): A Kagi Search API key
        query (str): The query to search for
        count (int): The number of results to return
    """
    url = 'https://kagi.com/api/v1/search'
    headers = {
        'Authorization': f'Bearer {api_key}',
    }
    params = {'query': query, 'limit': count}

    response = requests.post(url, headers=headers, json=params)
    response.raise_for_status()
    json_response = response.json()
    search_results = json_response.get('data', {}).get('search', [])

    results = [
        SearchResult(link=result['url'], title=result['title'], snippet=result.get('snippet'))
        for result in search_results
    ]

    print(results)

    if filter_list:
        results = get_filtered_results(results, filter_list)

    return results
```

**Reply by joey-zwaan** [reply]
> Even with access now granted, this doesn't work. Seems openwebui wants to use API v0, which doesn't seem to accept any of the access tokens I have, neither personal ones from https://kagi.com/settings/api nor the new ones from https://kagi.com/api/keys
> 
> For now, here's an edited `kagi.py` that supports the v1 API.
> 
> ```python
> import logging
> from typing import Optional
> 
> import requests
> from open_webui.retrieval.web.main import SearchResult, get_filtered_results
> 
> log = logging.getLogger(__name__)
> 
> 
> def search_kagi(api_key: str, query: str, count: int, filter_list: Optional[list[str]] = None) -> list[SearchResult]:
>     """Search using Kagi's Search API and return the results as a list of SearchResult objects.
> 
>     The Search API will inherit the settings in your account, including results personalization and snippet length.
> 
>     Args:
>         api_key (str): A Kagi Search API key
>         query (str): The query to search for
>         count (int): The number of results to return
>     """
>     url = 'https://kagi.com/api/v1/search'
>     headers = {
>         'Authorization': f'Bearer {api_key}',
>     }
>     params = {'query': query, 'limit': count}
> 
>     response = requests.post(url, headers=headers, json=params)
>     response.raise_for_status()
>     json_response = response.json()
>     search_results = json_response.get('data', {}).get('search', [])
> 
>     results = [
>         SearchResult(link=result['url'], title=result['title'], snippet=result.get('snippet'))
>         for result in search_results
>     ]
> 
>     print(results)
> 
>     if filter_list:
>         results = get_filtered_results(results, filter_list)
> 
>     return results
> ```
This works!
Thank you, you are a lifesaver.

Hope the team can use the fix in a new version?
